lottie/src/main/java/com/airbnb/lottie/utils/OffscreenLayer.java

```diff
@@ -0,0 +1,494 @@
+  }
+
+  private RenderStrategy chooseRenderStrategy(Canvas parentCanvas, RectF bounds, ComposeOp op) {
+    if (op.isNoop()) {
+      // Can draw directly onto the final canvas, results will look the same.
+      return RenderStrategy.DIRECT;
+    }
+
+    if (!op.hasShadow()) {
+      // Canvas.saveLayer() supports alpha-compositing, blend modes, and color filters, which is
+      // sufficient for this case, and is faster than manually maintaining an off-screen bitmap.
+      // It is not clear if it's faster than RENDER_NODE and when, but this is what we've been
+      // doing prior to OffscreenLayer, so keep that behavior to be safe and avoid regressions.
+      return RenderStrategy.SAVE_LAYER;
+    }
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && parentCanvas.isHardwareAccelerated()) {
+      if (op.hasShadow() && Build.VERSION.SDK_INT <= Build.VERSION_CODES.S) {
+        // RenderEffect, which we need for shadows, was only introduced in S. This means that, pre-S,
+        // we have to do renderShadow() on a bitmap, like in the BITMAP case. However, since it's not
+        // possible to draw a RenderNode to a software-rendering canvas, there would be no way to get
+        // our RenderNode contents onto a bitmap where we can renderShadow(). Therefore, fall back to
+        // full bitmap mode.
+        return RenderStrategy.BITMAP;
+      }
+
+      return RenderStrategy.RENDER_NODE;
+    }
+
+    // Slowest path: render to a bitmap, add a shadow manually.
+    return RenderStrategy.BITMAP;
+  }
+
+  private Bitmap allocateBitmap(RectF bounds, Bitmap.Config cfg) {
+    // Add 5% to the width and height before allocating. This avoids repeated
+    // reallocations in the worst cases of a bitmap growing every frame, while
+    // still being relatively speedy to blit and operate on.
+    int width = (int)Math.ceil(bounds.width() * 1.05);
+    int height = (int)Math.ceil(bounds.height() * 1.05);
+    Bitmap bmp = Bitmap.createBitmap(width, height, cfg);
+    return bmp;
+  }
+
+  private void deallocateBitmap(Bitmap bitmap) {
+    bitmap.recycle();
+  }
+
+  private boolean needNewBitmap(@Nullable Bitmap bitmap, RectF bounds) {
+    if (bitmap == null) {
+      return true;
+    }
+
+    if (bounds.width() >= bitmap.getWidth() || bounds.height() >= bitmap.getHeight()) {
+      return true;
+    }
+
+    // If the required area has reduced in size considerably, trigger a reallocation, since
+    // we might be paying a large unnecessary penalty to work with a bitmap that big.
+    if (bounds.width() < bitmap.getWidth() * 0.75f || bounds.height() < bitmap.getHeight() * 0.75f) {
+      return true;
+    }
+
+    return false;
+  }
+
+  public Canvas start(Canvas parentCanvas, RectF bounds, ComposeOp op) {
+    if (this.parentCanvas != null) {
+      throw new RuntimeException("Cannot nest start() calls on a single OffscreenBitmap - call finish() first");
```

Comment:
```suggestion
      throw new IllegalStateException("Cannot nest start() calls on a single OffscreenBitmap - call finish() first");
```

RuntimeExceptions usually shouldn't be thrown directly. Their subclasses are more specific.